### PR TITLE
MeasureGui: Add MeasureDistance X, Y and Z components back

### LINKED ISSUE #13741: Measurement: Allow to project measures
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

The new measurement feature gives the value of the measure and nothing else. It is sometimes useful to determine the 3 components of a measure in the base coordinate system. It may be desirable to determine said components in an arbitrary LCS, or the projected value on an user-selected plane or line.

Commercial CAD programs often provide the Δx, Δy, and Δz values for a distance, as you can see on the Solidworks screenshot below, but FreeCAD could be more complete

image

### Full version info

```shell
OS: Ubuntu 23.10 (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.37063 (Git) AppImage
Build type: Release
Branch: main
Hash: 680792030fd664c94829e2c1b78b21788e8d9879
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: French/France (fr_FR)
Installed mods: 
  * BillOfMaterials-WB 0.0.7
  * CosmeticThread3D
  * OpenTheme 2024.4.20
  * Curves 0.6.35
  * CurvedShapes 1.0.9
  * sheetmetal 0.4.13
```

### Subproject(s) affected?

Core

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by hlorus**
As [requested](#13741), this adds back the ability to display the X, Y and Z components of a measured distance. This PR adds the following:
- DistanceX, DistanceY, DistanceZ properties on the MeasureDistance feature
- The previously removed PartGui::DimensionLinear nodekit (now in the MeasureGui namespace)
- ShowDelta property on the ViewProviderMeasureDistance which is set to false by default
- The three component dimensions with a value text and colored in the axis colors

To see the components the user currently has to create a distance constraint and manually set the view property "Show Delta" to true.

Screenshot from 2024-06-27 11-20-30
Screenshot from 2024-06-28 14-23-07

fixes #13741

### CODE REVIEW

**Comment by hlorus** [comment]
@chennes Ideally the components wouldn't be shown if they have a value of zero. Apparently openinventor should have the SoMultiSwitch node which would be suitable for this however it seems that node is not available in the openinventor version freecad uses. Do you know any details here?

**Comment by pierreporte** [comment]
Is it possible to add Δx, Δy, and Δz prefixes in the 3D labels for each component?

**Comment by pierreporte** [comment]
For improved readability, the components labels should have a background like the main measure.

**Comment by hlorus** [comment]
> For improved readability, the components labels should have a background like the main measure.

Yeah sure, it now looks like that:
Screenshot from 2024-06-28 18-13-37


What do you think?

**Comment by pierreporte** [comment]
It lacks separation between the symbol and the value. Adding an equal sign would be good enough.

The Δz component is unreadable with your background so something is really needed.

**Comment by pierreporte** [comment]
> To see the components the user currently has to create a distance constraint and manually set the view property "Show Delta" to true.

Is this planned to be addressed by this PR or with a new one later? In the future, it should be a checkbox in the task panel. Its default state should be a program setting. The task panel should also contain 3 fields to display the components below the main value when it is checked.

**Comment by hlorus** [comment]
> > To see the components the user currently has to create a distance constraint and manually set the view property "Show Delta" to true.
> 
> Is this planned to be addressed by this PR or with a new one later? In the future, it should be a checkbox in the task panel. Its default state should be a program setting. The task panel should also contain 3 fields to display the components below the main value when it is checked.
> 
> 

Not with this PR, there will first have to be general support for measure settings.

**Comment by pierreporte** [comment]
Then I think that `ShowDelta` should be true by default so that the existence of the feature is obvious. Especially if this is merged for 1.0. It would be legitimate because Solidworks and Alibre already show the components by default.

**Comment by pierreporte** [comment]
Should I create a new issue for the missing equal sign and background?

**Comment by FEA-eng** [comment]
> I will clean up that typo and merge.

What's the typo? If it's just a trivial misspelling then maybe it would be better to merge it now (so that it's available in the next weekly builds for additional testing) and I can send a PR correcting it right after that.

File: src/Mod/Measure/Gui/ViewProviderMeasureDistance.cpp [hunk 1]
```diff
@@ -186,12 +358,70 @@ ViewProviderMeasureDistance::ViewProviderMeasureDistance()
...
+
+    pDeltaDimensionSwitch = new SoSwitch();
+    pDeltaDimensionSwitch->ref();
+    pGlobalSeparator->addChild(pDeltaDimensionSwitch);
+
+    pDeltaDimensionSwitch->addChild(dimDeltaX);
+    pDeltaDimensionSwitch->addChild(dimDeltaY);
+    pDeltaDimensionSwitch->addChild(dimDeltaZ);
+
+    // This should alredy be touched in ViewProviderMeasureBase
```
Comment (chennes) [inline src/Mod/Measure/Gui/ViewProviderMeasureDistance.cpp:416]:
```suggestion
    // This should already be touched in ViewProviderMeasureBase
```

**Comment by chennes** [comment]
Note to self: don't merge until InventorAll.h is modified to include
```
#include <Inventor/engines/SoDecomposeVec3f.h>
```
otherwise this won't compile with PCH enabled.

**Comment by Murmele** [comment]
I don't see any possibility to enable showing the delta. Is there something special required?

grafik

```
OS: Windows 11 build 22621
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38555 (Git)
Build type: Release
Branch: main
Hash: 73cb5bafe0c32c178d9b594a6ca03124483d305f
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: English/United States (en_US)
```

**Comment by FEA-eng** [comment]
> I don't see any possibility to enable showing the delta. Is there something special required?

It's in the properties once you save the dimension: https://wiki.freecad.org/Std_Measure#Properties

**Comment by Murmele** [comment]
I saved the measurement, but still I don't see anything

grafik

**Comment by FEA-eng** [comment]
> I saved the measurement, but still I don't see anything

You have to select this dimension in the tree view. Then you will see its properties below.

**Comment by Murmele** [comment]
I tried but it is not selectable

**Comment by FEA-eng** [comment]
> I tried but it is not selectable

You may have to close the Measurement tool first.

**Comment by Murmele** [comment]
Thanks this works! Is it possible to have already in the interactive mode the delta, for example with a checkbox. If the user saves then, the delta option will then also applied.

grafik

**Comment by Murmele** [comment]
I found a small issue: https://github.com/FreeCAD/FreeCAD/issues/16213

**Pull request merged**